### PR TITLE
fix: fix RPC behavior when exception is thrown across boundary

### PR DESCRIPTION
**Pull request by ngeojiajun-deriv**
# Context
When an error is thrown by the listener, it is _never_ deal properly by either client or the server side. Resulting to the following problems depending on the exception thrown.
1. Any attempt to throw blessed objects (including `Myriad::Exception` implementers) would cause immediate crash which would cause the Future for processing the RPC calls cancelled.
2. Any attempt to throw perl errors (ie. `die`) would cause caller to get `undef` as result instead of exceptions getting thrown.
3. Exception handling code in RPC wrongly assume the presence of the `does` method, causing the (1) to happen

# Changes
**This PR contains breaking changes, please test it before merge**
1. Properly deal with non-serializable objects when dealing with exceptions
2. Explicitly cast the error message to be always string
3. `call_rpc` would rethrow all exceptions from the RPC server as `Myriad::Exception::RPC::RemoteException`
4. Updated tests to cover these edge case

### CODE REVIEW

File: lib/Myriad/Exception/Base.pm [hunk 1]
```diff
@@ -46,6 +46,28 @@ Returns the exception message as a string.
...
+    return 0 unless $role and ref $self;
+
+    my $klass = ref $self;
+    my $klass_ptr = Object::Pad::MOP::Class->try_for_class($klass);
+    return 0 unless defined $klass_ptr;
+
+    my %roles = map {($_->name) => 1} $klass_ptr->all_roles;
+    return $roles{$role} // 0;
+}
+
```
Comment (nael-binary) [inline lib/Myriad/Exception/Base.pm:70]: instead of using and implementing `does`
or use this `Myriad::Exception::Base::does($e, 'Myriad::Exception')`

we use isa and regex class name to start with Myriad::Exception
isn't this easier and more straight forward?

Comment (ngeojiajun-deriv) [reply lib/Myriad/Exception/Base.pm]: this is addressed in next commit

File: lib/Myriad/RPC/Client/Implementation/Memory.pm [hunk 2]
```diff
@@ -88,14 +88,21 @@ async method call_rpc ($service, $method, %args) {
...
+                unless exists $message->response->{response};
+        }
+
         return $message->response->{response};
     } catch ($e) {
         if ($e =~ /Timeout/) {
             $e  = Myriad::Exception::RPC::Timeout->new(reason => 'deadline is due');
         } else {
-            $e = Myriad::Exception::InternalError->new(reason => $e) unless blessed $e && $e->does('Myriad::Exception');
+            $e = Myriad::Exception::InternalError->new(reason => $e) unless (ref $e) =~ /^Myriad::Exception::/;
```
Comment (tom-binary) [inline lib/Myriad/RPC/Client/Implementation/Memory.pm:103]: There's no guarantee the class starts with `Myriad::Exception::`, so this check is wrong - we specifically want to check that the exception *implements the Myriad::Exception rôle*.

File: lib/Myriad/RPC/Client/Implementation/Memory.pm [hunk 3]
```diff
@@ -88,14 +88,21 @@ async method call_rpc ($service, $method, %args) {
             $self->loop->timeout_future(at => $deadline),
             $pending
         );
+
+        unless (exists $message->response->{message}) {
+            my $reason = $message->response->{error}{message} // "Unknown";
+            Myriad::Exception::RPC::RemoteException->throw(reason => "Remote exception is thrown: $reason")
```
Comment (tom-binary) [inline lib/Myriad/RPC/Client/Implementation/Memory.pm:94]: The logic here doesn't look right - we're checking for `message` and `response`, why split up the conditional like this?

File: lib/Myriad/Service/Implementation.pm [hunk 4]
```diff
@@ -317,7 +317,16 @@ async method load () {
                     });
                     await $self->rpc->reply_success($service_name, $message, $response);
                 } catch ($e) {
-                    $e = Myriad::Exception::InternalError->new(reason => $e) unless (blessed $e and $e->does('Myriad::Exception'));
+                    # Pass through the myriad exception, use indirect exports from Myriad::Exception::Base
+                    # for simplicity
+                    unless ((ref $e) =~ /^Myriad::Exception::/) {
```
Comment (tom-binary) [inline lib/Myriad/Service/Implementation.pm:322]: Again, cannot rely on the package name here.

File: lib/Myriad/Service/Implementation.pm [hunk 5]
```diff
@@ -317,7 +317,16 @@ async method load () {
...
+                        # Ensure that we are passing the exception reason only by means of scalar
```
Comment (tom-binary) [inline lib/Myriad/Service/Implementation.pm:323]: The comment here is a bit confused: exceptions are _already_ scalars. "Scalar" just means there's only one value, the actual value could be an object, a number, a string, `undef` - anything that isn't a vector.

File: t/RPC/exceptions.t [hunk 6]
```diff
@@ -0,0 +1,82 @@
+use strict;
+use warnings;
+
+use Test::More;
+use Test::Deep;
+use Test::Fatal;
+use Test::Myriad;
```
Comment (tom-binary) [inline t/RPC/exceptions.t:7]: We shouldn't need to do any of this - we have `use Myriad::Class` for the standard Perl features, and `use Test::Myriad` for test-specific details, better to have that pull in any test-related components (can be addressed in a separate branch though).

File: lib/Myriad/Service/Implementation.pm [hunk 7]
```diff
@@ -317,7 +317,16 @@ async method load () {
...
+                        # or it would cause issue with strongly typed clients.
+                        # Moreover, attempt to use blessed object would crash the listener
+                        $e = "$e" if ref $e;
+                        $e =~ s/(\s+)$//; # Drop the trailling spaces
+                        $e = Myriad::Exception::InternalError->new(reason => $e);
```
Comment (tom-binary) [inline lib/Myriad/Service/Implementation.pm:328]: The `reason` handling in the exception class itself should be responsible for ensuring that `reason` is a valid string, converting as necessary - not the caller.

**Review by tom-binary — COMMENTED** [review]
Main issue here is the class check - we shouldn't be doing string checks like this.

File: lib/Myriad/RPC/Client/Implementation/Memory.pm (hunk 3 above)
Comment (ngeojiajun-deriv) [reply lib/Myriad/RPC/Client/Implementation/Memory.pm:94]: Could you please clarify on "split up the conditional"?

File: lib/Myriad/Service/Implementation.pm (hunk 5 above)
Comment (ngeojiajun-deriv) [reply lib/Myriad/Service/Implementation.pm:323]: Oops, I meant the non-reference scalars because we expects the exceptions is something like strings but not HASHREFs. Will fix the comment here

File: t/RPC/exceptions.t (hunk 6 above)
Comment (ngeojiajun-deriv) [reply t/RPC/exceptions.t:7]: I have tried with `Myriad::Class` but it produce this error
 ```
Prototype mismatch: sub main::blessed ($) vs none at t/RPC/exceptions.t line 3.
```

File: lib/Myriad/Exception/Base.pm [hunk 8]
```diff
@@ -27,7 +27,15 @@ use overload '""' => sub { shift->as_string }, bool => sub { 1 }, fallback => 1;
...
     my ($class, %args) = @_;
-    bless \%args, $class
+    # Force the reason to be a valid string
+    # or it would cause issue with strongly typed clients.
+    # Moreover, attempt to use blessed object here would crash the listener
+    if ($args{reason}) {
+        my $reason = "$args{reason}";
+        $reason =~ s/(\s+)$//;
+        $args{reason} = $reason;
+    }
```
Comment (nael-binary) [inline lib/Myriad/Exception/Base.pm:37]: whats the motivation behind this change?
did we ever have issue where we are passing other than string in reason?

File: lib/Myriad/Exception/Base.pm (hunk 1 above)
Comment (nael-binary) [reply lib/Myriad/Exception/Base.pm:78]: I don't understand the need of having this?

File: lib/Myriad/RPC/Client/Implementation/Memory.pm [hunk 9]
```diff
@@ -88,14 +88,20 @@ async method call_rpc ($service, $method, %args) {
...
+            Myriad::Exception::RPC::RemoteException->throw(reason => "Remote exception is thrown: $reason");
+        }
+
         return $message->response->{response};
     } catch ($e) {
         if ($e =~ /Timeout/) {
             $e  = Myriad::Exception::RPC::Timeout->new(reason => 'deadline is due');
         } else {
-            $e = Myriad::Exception::InternalError->new(reason => $e) unless blessed $e && $e->does('Myriad::Exception');
+            $e = Myriad::Exception::InternalError->new(reason => $e) unless Myriad::Exception::Base::does($e, 'Myriad::Exception');
```
Comment (nael-binary) [inline lib/Myriad/RPC/Client/Implementation/Memory.pm:102]: can't we use `isa` here?

File: lib/Myriad/Exception/Base.pm (hunk 8 above)
Comment (ngeojiajun-deriv) [reply lib/Myriad/Exception/Base.pm:37]: This is to prevent something like `HTTP::Exception` to be passed over RPC boundary. Or even something like ARRAYREF and HASHREF can cause portability issue if we implement Myriad in other languages.

File: lib/Myriad/RPC/Client/Implementation/Redis.pm [hunk 10]
```diff
@@ -95,14 +95,19 @@ async method call_rpc($service, $method, %args) {
         # The subscription loop will parse the message for us
         my $message = await Future->wait_any($self->loop->timeout_future(after => $timeout), $pending);
 
+        unless (exists $message->response->{response}) {
+            my $reason = $message->response->{error}{message} // "Unknown";
+            Myriad::Exception::RPC::RemoteException->throw(reason => "Remote exception is thrown: $reason");
+        }
+
```
[the remark this answers is not in the document]
Comment (ngeojiajun-deriv) [reply lib/Myriad/RPC/Client/Implementation/Redis.pm:102]: The thing here is the `$message` is considered valid but it indicates remote failure. So IMO it is better to put the failure check as near to caller when ever possible.

File: lib/Myriad/Exception/Base.pm (hunk 8 above)
Comment (tom-binary) [inline lib/Myriad/Exception/Base.pm:37]: I'd suggest either `trim` or just do the regex directly, no need to assign to other variables - regex s// operation already coërces to string:

```suggestion
    $args{reason} =~ s{\s+$}{} if defined $args{reason};
```

**Comment by tom-binary** [comment]
Since there's not been any progress on this for a few months, I've merged #353 for basic exception support.
